packages/compass-import-export/src/utils/csv.ts

```diff
@@ -0,0 +1,310 @@
+    }
+
+    let type = included[fieldName];
+    if (type === 'mixed') {
+      type = detectFieldType(original, ignoreEmptyStrings);
+    }
+
+    const path = parsedHeader[name];
+
+    try {
+      const value = parseValue(original, type);
+
+      placeValue(doc, path, value, true);
+    } catch (err: unknown) {
+      // rethrow with the column index appended to aid debugging
+      (err as Error).message = `${(err as Error).message} [Col ${index}]`;
+      throw err;
+    }
+  }
+
+  return doc;
+}
+
+export function parseValue(value: string, type: CSVFieldType): CSVValue {
+  if (type === 'int') {
+    return new Int32(value);
+  }
+
+  if (type === 'long') {
+    return new Long(value);
+  }
+
+  if (type === 'double') {
+    return new Double(parseFloat(value));
+  }
+
+  if (type === 'boolean') {
+    if (value === 'true' || value === 'TRUE') {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  if (type === 'date') {
+    return new Date(value);
+  }
+
+  if (type === 'null') {
+    // This will only match an explicit 'Null' in the CSV file because empty
+    // strings are handled separately already.
+    return null;
+  }
+
+  // By default leave it as a string
+  return value;
+}
+
+export function parseHeaderName(value: string): PathPart[] {
+  const parts: PathPart[] = [];
+
+  let previousType: 'field' | 'index' = 'field';
+  let type: 'field' | 'index' = 'field';
+  let snippet: string[] = [];
+
+  for (const char of value) {
+    if (type === 'field') {
+      if (char === '[') {
+        // this snippet length check is for:
+        // 1. nested arrays (because closing an array default to type field)
+        // 2. top-level array paths like [0].foo
+        if (snippet.length) {
+          parts.push({ type: 'field', name: snippet.join('') });
+        } else if (previousType === 'field') {
+          // this supports the edge case where the field name is a blank string
+          // at the top level that is the name of an array field.
+          parts.push({ type: 'field', name: '' });
+        }
+        previousType = type;
+        type = 'index';
+        snippet = [];
+        continue;
+      }
+      if (char === '.') {
+        if (snippet.length) {
+          // this snippet length check helps with arrays of objects like
+          // array[2].foo. closing the array defaults to type field and then
+          // immediately afterwards we encounter a .
+          parts.push({ type: 'field', name: snippet.join('') });
+        } else if (previousType === 'field') {
+          // this supports the edge case where the field name is a blank string
+          // inside an object or a blank string at the top level that is the
+          // name of an object.field.
+          parts.push({ type: 'field', name: '' });
+        }
+        snippet = [];
+        continue;
+      }
+    } else {
+      if (char === ']') {
+        parts.push({ type: 'index', index: parseInt(snippet.join(''), 10) });
```

Comment:
Maybe rather

```suggestion
        parts.push({ type: 'index', index: +snippet.join('') });
```

+ some handling for `NaN` results? `parseInt()` cuts off processing at the first invalid character, if there are any, so that’s probably something we should handle here, and some invalid strings would result in `NaN` with either conversion method

Reply:
I'll add some unit tests for this function. Would have to throw and catch somewhere, make that kind of thing cause the header parsing to fail because there's not much we can do to recover from broken header strings like that.